File: go-sidecar/robotics/passport.go
// Scannable robot passport (Phase 5.6), Go.
//
// Mirrors vouch/robotics/passport.py and the TypeScript SDK. A compact, signed
// RobotPassport that anyone can scan (QR or NFC) to check a robot's owner,
// authorized actions, certification, and current standing, offline. The QR/NFC
// payload is a vouch-passport: URI carrying the multibase JCS bytes of the
// credential, so an offline reader verifies the signature with no network call.
// The URI encoding is deterministic, so a passport encoded by any language
// decodes and verifies in the others.
package robotics

import (
	"crypto/ed25519"
	"encoding/json"
	"strings"
	"time"

	"github.com/vouch-protocol/vouch/go-sidecar/signer"
)

// Passport constants.
const (
	RobotPassportType    = "RobotPassport"
	PassportURIScheme    = "vouch-passport:"
	StatusActive         = "active"
	StatusSuspended      = "suspended"
	StatusDecommissioned = "decommissioned"
)

// PassportError is returned for malformed passport URIs.
type PassportError struct{ Msg string }

func (e *PassportError) Error() string { return e.Msg }

// BuildPassportOptions configures BuildPassport.
type BuildPassportOptions struct {
	RobotDID          string
	Make              string
	Model             string
	Owner             string
	AuthorizedActions []string
	Certification     string // "" omits the field
	Status            string // "" defaults to active
	ValidSeconds      int    // 0 omits validUntil
	ValidFrom         time.Time
}

// BuildPassport builds a signed RobotPassport credential (issued by the robot or
// an authority).
func BuildPassport(s *signer.Signer, opts BuildPassportOptions) (map[string]any, error) {
	issued := opts.ValidFrom
	if issued.IsZero() {
		issued = time.Now().UTC()
	}
	status := opts.Status
	if status == "" {
		status = StatusActive
	}
	subject := map[string]any{
		"id":                opts.RobotDID,
		"make":              opts.Make,
		"model":             opts.Model,
		"owner":             opts.Owner,
		"authorizedActions": strsToAny(opts.AuthorizedActions),
		"status":            status,
	}
	if opts.Certification != "" {
		subject["certification"] = opts.Certification
	}
	cred := map[string]any{
		"@context":          []any{vcContextV2, vouchContextV1},
		"type":              []any{"VerifiableCredential", RobotPassportType},
		"issuer":            s.DID(),
		"validFrom":         iso(issued),
		"credentialSubject": subject,
	}
	if opts.ValidSeconds > 0 {
		cred["validUntil"] = iso(issued.Add(time.Duration(opts.ValidSeconds) * time.Second))
	}
	return s.AttachProof(cred)
}

// EncodePassport encodes a passport into a compact vouch-passport: URI for a QR
// or NFC tag.
func EncodePassport(passport map[string]any) (string, error) {
	canon, err := signer.Canonicalize(passport)
	if err != nil {
		return "", err
	}
	return PassportURIScheme + mb64(canon), nil
}

// DecodePassport decodes a vouch-passport: URI back into the passport credential.
func DecodePassport(uri string) (map[string]any, error) {
	if !strings.HasPrefix(uri, PassportURIScheme) {
		return nil, &PassportError{"not a " + PassportURIScheme + " URI"}
	}
	body := uri[len(PassportURIScheme):]
	raw, err := unmb64(body)
	if err != nil {
		return nil, &PassportError{"expected multibase 'u' payload"}
	}
	var p map[string]any
	if err := json.Unmarshal(raw, &p); err != nil {
		return nil, &PassportError{"passport decode failed: " + err.Error()}
	}
	return p, nil
}

// PassportSummary is the human-readable result of a successful verification.
type PassportSummary struct {
	Robot             string
	Make              string
	Model             string
	Owner             string
	AuthorizedActions []string
	Certification     string
	Status            string
}

// VerifyPassport verifies a passport credential object. A suspended or
// decommissioned status still verifies but is surfaced in the summary so a
// scanner can refuse cooperation; an expired passport fails. A zero now uses the
// current time.
func VerifyPassport(passport map[string]any, publicKey ed25519.PublicKey, now time.Time) (bool, *PassportSummary) {
	if !hasType(passport["type"], RobotPassportType) {
		return false, nil
	}
	if ok, err := signer.VerifyDataIntegrityProof(passport, publicKey); err != nil || !ok {
		return false, nil
	}
	if now.IsZero() {
		now = time.Now().UTC()
	}
	if vu, _ := passport["validUntil"].(string); vu != "" {
		if t, err := time.Parse(time.RFC3339, vu); err == nil && now.After(t) {
			return false, nil
		}
	}

	s, _ := passport["credentialSubject"].(map[string]any)
	robot, _ := s["id"].(string)
	mk, _ := s["make"].(string)
	model, _ := s["model"].(string)
	owner, _ := s["owner"].(string)
	cert, _ := s["certification"].(string)
	status, _ := s["status"].(string)
	return true, &PassportSummary{
		Robot:             robot,
		Make:              mk,
		Model:             model,
		Owner:             owner,
		AuthorizedActions: toStrSlice(s["authorizedActions"]),
		Certification:     cert,
		Status:            status,
	}
}

// VerifyPassportURI decodes a vouch-passport: URI and verifies it. Returns
// (false, nil) if the URI is malformed.
func VerifyPassportURI(uri string, publicKey ed25519.PublicKey, now time.Time) (bool, *PassportSummary) {
	p, err := DecodePassport(uri)
	if err != nil {
		return false, nil
	}
	return VerifyPassport(p, publicKey, now)
}

File: packages/sdk-ts/src/robotics/passport.ts
/**
 * Scannable robot passport (Phase 5.6), TypeScript.
 *
 * Mirrors `vouch/robotics/passport.py`. A compact, signed RobotPassport that
 * anyone can scan (QR or NFC) to check a robot's owner, authorized actions,
 * certification, and current standing, offline. The QR/NFC payload is a
 * `vouch-passport:` URI carrying the multibase JCS bytes of the credential, so
 * an offline reader verifies the signature with no network round-trip. The URI
 * encoding is deterministic, so a passport encoded by either language decodes
 * and verifies in the other.
 */

import * as crypto from 'crypto';

import { verifyProof } from '../data-integrity';
import { canonicalize } from '../jcs';
import type { Signer } from '../signer';
import { VC_CONTEXT_V2, VOUCH_CONTEXT_V1 } from '../vc';

export const ROBOT_PASSPORT_TYPE = 'RobotPassport';
export const PASSPORT_URI_SCHEME = 'vouch-passport:';
export const STATUS_ACTIVE = 'active';
export const STATUS_SUSPENDED = 'suspended';
export const STATUS_DECOMMISSIONED = 'decommissioned';

export class PassportError extends Error {}

function iso(d: Date): string {
  return d.toISOString().replace(/\.\d{3}Z$/, 'Z');
}

export interface BuildPassportOptions {
  robotDid: string;
  make: string;
  model: string;
  owner: string;
  authorizedActions: string[];
  certification?: string;
  status?: string;
  validSeconds?: number;
  validFrom?: Date;
}

/** Build a signed RobotPassport credential (issued by the robot or an authority). */
export async function buildPassport(
  signer: Signer,
  opts: BuildPassportOptions
): Promise<Record<string, unknown>> {
  const issued = opts.validFrom ?? new Date();
  const subject: Record<string, unknown> = {
    id: opts.robotDid,
    make: opts.make,
    model: opts.model,
    owner: opts.owner,
    authorizedActions: [...opts.authorizedActions],
    status: opts.status ?? STATUS_ACTIVE,
  };
  if (opts.certification !== undefined) subject.certification = opts.certification;

  const credential: Record<string, unknown> = {
    '@context': [VC_CONTEXT_V2, VOUCH_CONTEXT_V1],
    type: ['VerifiableCredential', ROBOT_PASSPORT_TYPE],
    issuer: signer.getDid(),
    validFrom: iso(issued),
    credentialSubject: subject,
  };
  if (opts.validSeconds !== undefined) {
    credential.validUntil = iso(new Date(issued.getTime() + opts.validSeconds * 1000));
  }
  return signer.attachProof(credential);
}

/** Encode a passport into a compact vouch-passport: URI for a QR or NFC tag. */
export function encodePassport(passport: Record<string, unknown>): string {
  const blob = Buffer.from(canonicalize(passport)).toString('base64url');
  return PASSPORT_URI_SCHEME + 'u' + blob;
}

/** Decode a vouch-passport: URI back into the passport credential. */
export function decodePassport(uri: string): Record<string, unknown> {
  if (!uri.startsWith(PASSPORT_URI_SCHEME)) throw new PassportError(`not a ${PASSPORT_URI_SCHEME} URI`);
  const body = uri.slice(PASSPORT_URI_SCHEME.length);
  if (!body.startsWith('u')) throw new PassportError("expected multibase 'u' payload");
  const raw = Buffer.from(body.slice(1), 'base64url');
  return JSON.parse(raw.toString('utf8'));
}

export interface PassportSummary {
  robot?: string;
  make?: string;
  model?: string;
  owner?: string;
  authorizedActions: string[];
  certification?: string;
  status?: string;
}

/**
 * Verify a passport (a credential object or a vouch-passport: URI). A suspended
 * or decommissioned status still verifies but is surfaced in the summary so a
 * scanner can refuse cooperation. An expired passport fails.
 */
export function verifyPassport(
  passport: Record<string, any> | string,
  publicKey: crypto.KeyObject,
  opts: { now?: Date } = {}
): { ok: boolean; summary?: PassportSummary } {
  let p: Record<string, any>;
  if (typeof passport === 'string') {
    try {
      p = decodePassport(passport);
    } catch {
      return { ok: false };
    }
  } else {
    p = passport;
  }

  const types = Array.isArray(p.type) ? p.type : [p.type];
  if (!types.includes(ROBOT_PASSPORT_TYPE)) return { ok: false };
  try {
    if (!verifyProof(p, publicKey)) return { ok: false };
  } catch {
    return { ok: false };
  }

  const now = opts.now ?? new Date();
  if (p.validUntil) {
    const vu = new Date(p.validUntil);
    if (!Number.isNaN(vu.getTime()) && now > vu) return { ok: false };
  }

  const s = (p.credentialSubject ?? {}) as Record<string, any>;
  return {
    ok: true,
    summary: {
      robot: s.id,
      make: s.make,
      model: s.model,
      owner: s.owner,
      authorizedActions: s.authorizedActions ?? [],
      certification: s.certification,
      status: s.status,
    },
  };
}
